src/main/java/com/advent/of/code/day5/VentsRadar.java

```diff
@@ -0,0 +1,29 @@
+package com.advent.of.code.day5;
+
+import java.util.List;
+
+class VentsRadar {
+    long getIntersectionsOf(List<Segment> segments, int atLeast) {
+        int xmin = segments.stream().mapToInt(Segment::getMinX).min().getAsInt();
+        int ymin = segments.stream().mapToInt(Segment::getMinY).min().getAsInt();
+        int xmax = segments.stream().mapToInt(Segment::getMaxX).max().getAsInt();
+        int ymax = segments.stream().mapToInt(Segment::getMaxY).max().getAsInt();
+
+        long result = 0L;
+
+        for (int i = xmin; i <= xmax; i++) {
+            for (int j = ymin; j <= ymax; j++) {
+                long intersections = 0;
+                for (Segment segment : segments) {
```

Comment:
zwróć tutaj uwagę na złożoność O(n^3) - przez to, że mamy w sobie zagnieżdżone aż trzy pętle (które iterują po n [lub nawet więcej bo i oraz j należą do większych przedziałów] elementów gdzie n to ilość segmentów) ten fragment to niezły bottle neck - faktycznie test dla zadania drugiego wykonuje się aż 4,5s (wielokrotnie dłużej niż reszta testów)

tego raczej przyspieszyć się nie da bo głównym kosztem nie jest operacja wykonywana w pętli tylko przejście po tych wszystkich elementach - tutaj można się wspomóc tylko jakimś cwanym algorytmem (istnieją np algorytmy zamiatania związane z otoczką wypukłą w przestrzeni, które by "pogrupowały" nam wstępnie odcinki na te które mogą się przeciąć albo nie przetną się nigdy i pozwolą takie zadanie zrobić w O(logn) czyli znacznie szybciej :) 

https://wazniak.mimuw.edu.pl/index.php?title=Zaawansowane_algorytmy_i_struktury_danych/Wyk%C5%82ad_12

no ale powiedzmy że 4s to nie jest jakaś masakra dlatego nie bawiłem się w to, zwłaszcza że nie umiem napisać takiego algorytmu z głowy i musiałbym albo spędzić sporo czasu na rozkminianiu go albo skopiować skądś :P ale wiem, że istnieje

Reply:
Z tym grupowaniem to mi przyszło coś takiego do głowy w mojej klasie Line, ponieważ oprócz pól określających początek (x1,y1) i koniec (x2,y2) to jeszcze pole shape, które mogło przyjąć jedną z wartości enuma Shape: HORIZONTAL, VERTICAL, DIAGONAL. Teoretycznie miało to pomóc w późniejszym sprawdzaniu punktów przecięcia tj. jeżeli dwie linie są HORIZONTAL to porównywałem czy ich "y" są równe: jeśli tak to szukamy punktów, w przeciwnym wypadku przeskakiwałem dalej itd. W teorii wszystko było ok, ale nie mogłem uzyskać poprawnego wyniku, więc zrezygnowany zacząłem szukać podpowiedzi na youtube :)